components/system-broker/internal/director/client_test.go

```diff
@@ -0,0 +1,860 @@
+					fieldRegex := regexp.MustCompile(`\b` + expectedProp + `\b`)
+
+					matches := fieldRegex.FindAllStringIndex(query, -1)
+					actualCount := len(matches)
+
+					assert.Equal(t, expectedCount, actualCount, expectedProp)
+				}
+			}
+		})
+	}
+}
+
+func TestGraphQLClient_RequestPackageInstanceCredentialsCreation(t *testing.T) {
+	type fields struct {
+		getGCLI           func() *directorfakes.FakeClient
+		inputGraphqlizer  director.GraphQLizer
+		outputGraphqlizer director.GqlFieldsProvider
+	}
+	type testCase struct {
+		name             string
+		fields           fields
+		credentialsInput *director.PackageInstanceCredentialsInput
+		expectedErr      string
+		expectedQuery    string
+	}
+
+	inputGraphqlizer := &graphqlizer.Graphqlizer{}
+	outputGraphqlizer := &graphqlizer.GqlFieldsProvider{}
+
+	tests := []testCase{
+		{
+			name: "success",
+			fields: fields{
+				getGCLI: func() *directorfakes.FakeClient {
+					fakeGCLI := &directorfakes.FakeClient{}
+					fakeGCLI.DoReturns(nil)
+					return fakeGCLI
+				},
+				inputGraphqlizer:  inputGraphqlizer,
+				outputGraphqlizer: outputGraphqlizer,
+			},
+			credentialsInput: &director.PackageInstanceCredentialsInput{
+				PackageID: "packageID",
+				AuthID:    "authID",
+			},
+			expectedQuery: "mutation {\n\t\t\t  result: requestPackageInstanceAuthCreation(\n\t\t\t\tpackageID: \"packageID\"\n\t\t\t\tin: {\n\t\t\t\t  id: \"authID\"\n\t\t\t\t  context: \"null\"\n    \t\t\t  inputParams: \"null\"\n\t\t\t\t}\n\t\t\t  ) {\n\t\t\t\t\tstatus {\n\t\t\t\t\t  condition\n\t\t\t\t\t  timestamp\n\t\t\t\t\t  message\n\t\t\t\t\t  reason\n\t\t\t\t\t}\n\t\t\t  \t }\n\t\t\t\t}",
+		},
+		{
+			name: "when gql client returns an error",
+			fields: fields{
+				getGCLI: func() *directorfakes.FakeClient {
+					fakeGCLI := &directorfakes.FakeClient{}
+					fakeGCLI.DoReturns(errors.New("some error"))
+					return fakeGCLI
+				},
+				inputGraphqlizer:  inputGraphqlizer,
+				outputGraphqlizer: outputGraphqlizer,
+			},
+			credentialsInput: &director.PackageInstanceCredentialsInput{
+				PackageID: "packageID",
+				AuthID:    "authID",
+			},
+			expectedErr: "while executing GraphQL call to create package instance auth: some error",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			gcli := tt.fields.getGCLI()
+			c := director.NewGraphQLClient(
+				gcli,
+				tt.fields.inputGraphqlizer,
+				tt.fields.outputGraphqlizer,
+			)
+			_, err := c.RequestPackageInstanceCredentialsCreation(context.TODO(), tt.credentialsInput)
+			if tt.expectedErr != "" {
+				assert.Error(t, err)
+				assert.Contains(t, err.Error(), tt.expectedErr)
+			} else {
+				assert.Equal(t, 1, gcli.DoCallCount())
+
+				_, graphqlReq, _ := gcli.DoArgsForCall(0)
+				query := graphqlReq.Query()
+				assert.Equal(t, tt.expectedQuery, query)
+			}
+		})
+	}
+}
+
+func TestGraphQLClient_FetchPackageInstanceCredentials(t *testing.T) {
+	type fields struct {
```

Comment:
it will be almost the same as the function will have to accept almost all the different parameters that are in the struct currently. There are test cases that need different things. I think it is clearer this way.

Reply:
you could make all the funcs use fakeGCLI.DoStub, remove the t *testing.T arg, create a client
```
	client := func(gliResponse string, err error) *directorfakes.FakeClient {
		fakeGCLI := &directorfakes.FakeClient{}
		fakeGCLI.DoStub = func(c context.Context, g *graphql.Request, i interface{}) error {
			if err != nil {
				return err
			}
			err := json.Unmarshal([]byte(gliResponse), i)
			assert.NoError(t, err)
			return nil
		}
		return fakeGCLI
	}
```

i think test cases would like slightly simpler 
![image](https://user-images.githubusercontent.com/8989462/101938452-02c94e80-3bec-11eb-91a9-0a61a2345cf2.png)

Reply:
agree with that, I thought it was about the whole test case struct